### PR TITLE
Preprocess dominator tree to answer queries in O(1)

### PR DESCRIPTION
**Pull request by tmiasko**
r? @ghost

### CODE REVIEW

File: compiler/rustc_data_structures/src/graph/dominators/mod.rs [hunk 1]
```diff
@@ -321,3 +309,101 @@ impl<'dom, Node: Idx> Iterator for Iter<'dom, Node> {
...
+            edges[idom].end += 1;
+        }
+    }
+    let mut m = 0;
+    for e in edges.iter_mut() {
+        m += e.end;
+        e.start = m;
+        e.end = m;
+    }
+    let mut node = IndexVec::from_elem_n(Idx::new(0), m.try_into().unwrap());
```
Comment (bryangarza) [inline compiler/rustc_data_structures/src/graph/dominators/mod.rs:373]: if you make the `m` a `usize` from the start, you can avoid the `try_into().unwrap()`

**Review by bryangarza — CHANGES_REQUESTED** [review]
I'm not an official reviewer, but wanted to leave some feedback.

Just had one comment in the code itself, other than that, it'd be nice to see a unit test or two for the `dominators` function before merging.

**Comment by pnkfelix** [comment]
@tmiasko you wrote in a separate comment: https://github.com/rust-lang/rust/pull/107449#issuecomment-1426996371

> FYI: https://github.com/rust-lang/rust/pull/107157 is intended for instrument coverage builds, where the cost of dominance queries often exceeds the cost of initial preprocessing - the opposite of a profile here, where at best it wouldn't be a regression.

I'm trying to connect that statement with the PR here.

Is the preprocessing done here supposed to be conditionally enabled? I haven't yet seen that in this PR, but I only did a quick skim so maybe I overlooked it.

**Comment by tmiasko** [comment]
> Is the preprocessing done here supposed to be conditionally enabled?

The extra cost seems negligible, so it can be unconditional (one can't quite tell from last perf results alone, since the measurement is biased by a regression in obligation processing, which has unstable codegen for some reason).

This pull request isn't quite ready for review (hence no assigned reviewer). I also wanted to add a new benchmark that demonstrates perf aspect addressed here first.

Closing for now, maybe will return to this later.

**Pull request closed without merging**